Implement WAN protocol

mmedenjak opened the pull request:
Added support for setting, transmitting and reading the negotiated WAN
protocol version during the serialisation/deserialization process. We
reuse the existing version field and bytes used for intra-cluster
versioning (RU) but use negative values to denote WAN protocol version.
As such, it is not possible to set both the WAN protocol version and
cluster version for RU on a single versioned instance, which is not a
real restriction as the instance will either be sent to a cluster member
or a WAN target, not both at the same time.

Also did some minor refactoring such as moved some methods and extracted
inner classes.

**Note to reviewers**
Moved the `BufferObjectDataOutput#setVersion` and `BufferObjectDataInput#setVersion` methods to `VersionAware` and created a corresponding `WanProtocolVersionAware`. Tried to rename the `VersionAware` to `Versioned` then since it has the setter but then it clashes with `com.hazelcast.nio.serialization.impl.Versioned` so leaved it as-is. Also added `VersionedObjectDataOutput#getRawVersion` and `VersionedObjectDataInput#getRawVersion` which is not present on any interface. I'm open to suggestions for improvements about naming and placement of methods.

EE: https://github.com/hazelcast/hazelcast-enterprise/pull/3104

blazember commented on hazelcast/src/test/java/com/hazelcast/version/VersionTest.java:
```diff
@@ -41,44 +42,49 @@
         assertFalse(V3_0.isGreaterThan(of(3, 0)));
         assertFalse(V3_0.isGreaterThan(of(4, 0)));
+        assertTrue(V3_0.isGreaterThan(NEGATIVE_VERSION_1));
+        assertTrue(NEGATIVE_VERSION_1.isGreaterThan(NEGATIVE_VERSION_2));
```
I'd expect this to be the other way around, when version is a WAN version.

I'm thinking about subclassing `Version`. That way we can create `ClusterVersion` and `WanVersion` that can
- compare/sort the versions properly,
- validate that once we set a cluster version, it can't be replaced by a WAN version (checking the sign bit) and vice-versa.

WDYT?

mmedenjak replied:
Although we might argue that you need to know that negative values are also used as a special case for another kind of version to say that -2 should be greater than -1, I do agree this might be a bit confusing. Also, the fact that there is negative versions in the first place is confusing. I'll probably change this to only allow positive versions and restrict the negative version magic to only serialisation/deserialisation.

blazember commented on hazelcast/src/main/java/com/hazelcast/nio/WanProtocolVersionAware.java:
```diff
@@ -0,0 +1,46 @@
+     * {@link VersionAware#getVersion()} but is used in WAN replication, not in
+     * message exchange inside a single cluster.
+     *
+     * @return the WAN protocol version or {@link Version#UNKNOWN} if not set
```
If the version set is a cluster version, shouldn't it rather be an exception? Same for the other way around. Getting `UNKNOWN` in these cases suggests that the version isn't understood rather than that it's not interpreted and it lets the caller continue based on an invalid state.

mmedenjak replied:
Yes, you might be right. I'll check this out.

vbekiaris commented:
In general looks good. The removal of non-negative assertions from `Version` constructor seems like a potential issue, eg I can now do `setWanProtocolVersion(Version.of(-4, 0)` --> `isWanProtocolVersionSet() -> false`. Either introducing `WanVersion` subclass for cleaner separation or keeping existing API with stricter validation seem ok for me.

Having in mind that negative major version is an internal implementation detail, I don't think negative major version comparison is an issue since these should never leak outside the `VersionedObjectDataInput/Output`.

mmedenjak commented:
@vbekiaris @blazember addressed review comments in a separate commit and rebased on top of latest master.
- `Version` can no longer be negative
- there are now two versions in `VersionedObjectDataInput` and `VersionedObjectDataOutput`. This simplifies the getters and setters but causes both versions to need to be propagated to other internal API.
- the negative version trick is now limited to serialisation/deserialisation in `EnterpriseDataSerializableSerializer#readVersions` and `EnterpriseDataSerializableSerializer#writeIdentifiedDataSerializable` and `EnterpriseDataSerializableSerializer#writeDataSerializable`.
- we still don't throw an exception when a WAN related class is used in WAN context or vice-versa. This can be added but only if the object header actually carries version info. If there is no serialised version info, we don't know for what the class is being used. If you think we need to throw the exception, please say so. For now, I don't think it's necessary.

blazember commented:
LGTM :+1: 

Maybe we could hide the WAN protocol version from the non-WAN operations by making `VersionedObjectDataInput|Output` not to implement `WanProtocolVersionAware`. Instead, we can "convert" it with a `asWanProtocolVersionAware(VersionedObjectDataInput|Output)` helper method. That could either return an instance that implements only the `WanProtocolVersionAware` interface, or one that delegates to the `VersionedObjectDataInput|Output` instance plus implements `WanProtocolVersionAware`. Not confident it makes the effort though, so I leave it only just an idea.

blazember commented:
run-lab-run

blazember commented:
run-lab-run

Holmistr commented:
run-lab-run

mmedenjak commented:
Thank you for the reviews @vbekiaris @blazember ! To get as much done for feature freeze, I'll keep the WAN protocol code as-is and open a new issue for future enhancements, such as the one with `WanProtocolVersionAware`.
